### DISCUSSION TITLE
Validation doesn't run when calling trigger right after reset

### DISCUSSION THREAD

**Discussion by Kadrian** [react-hook-form/react-hook-form#10823] [Q&A] [answered]
Hi! I'm working on a form that wants to do an initial `.trigger()` right away, to see errors right from the start.

Now when resetting the form, I'm doing something like this:
```javascript
methods.reset({ ...methods.getValues(), ...defaultValues });
methods.trigger(); // <-- won't work / won't trigger the validation
```

while validation will run and error are showing like this: 
```javascript
methods.reset({ ...methods.getValues(), ...defaultValues });
setTimeout(() => methods.trigger(), 0); // <-- works
```

is that known / intentional? Is `methods.reset` doing something async? What could I do instead to avoid the `setTimeout`?

**Comment by Moshyfawn** [comment, marked as answer]
It's a known limitation that was partly addressed in issue #9863.

Regarding your specific use case, there are two aspects to consider:

- Firstly, given that your example demonstrates the propagation of user input into the `reset` values, you can utilize the `keepValues` property of the `reset` method to retain the previously entered values:

```js
reset(defaultValues, {
  keepValues: true
});
```

- Secondly, if it's necessary to validate the user input upon form reset, you can subscribe to the `isDirty` state of the form. This state should reset along with the form, as you're providing new default values for the form.

Check out [this codesandbox](https://codesandbox.io/p/sandbox/dazzling-meitner-nw4htd?file=%2Fsrc%2FApp.tsx%3A7%2C11) for example.

```jsx
import { useEffect } from "react"
import { useForm } from "react-hook-form"

const defaultValues = {
  name: ""
}

function App() {
  const { formState, handleSubmit, register, reset, trigger } = useForm({
    defaultValues
  })

  const { errors, isDirty } = formState

  useEffect(() => {
    if (!isDirty) {
      trigger()
    }
  }, [isDirty, trigger])

  return (
    <form onSubmit={handleSubmit(console.log)}>
      <label>
        Name
        <input
          {...register("name", {
            required: "REQUIRED",
            minLength: { value: 10, message: "MIN 10 CHAR" }
          })}
        />
      </label>
      {errors.name && <p role="alert">{errors.name.message}</p>}
      <button
        type="button"
        onClick={() => {
          reset(defaultValues, {
            keepValues: true
          })
        }}
      >
        Reset
      </button>
      <button type="submit">Submit</button>
    </form>
  )
}

export default App
```

**Reply by Kadrian** [reply]
Thank you for all the explanation!

**Comment by tmerve** [comment]
Would checking isReady still work with a separate useEffect, to only check initial render?

```tsx
useEffect(() => {
  if (isReady) {
    form.trigger();
  }
}, [form, isReady]);
```

and what if we already reset values with `useEffect` like this?

```tsx
  useEffect(() => {
    form.reset(defaultValues);
  }, [form, defaultValues]);
```

in such a case, even if they work fine, two useEffects seems a bit redundant
